### PR TITLE
add Dockerfile

### PR DESCRIPTION
**Pull request by phstudy**
Dockerfile is a good way to help users to test functionality of the extension without manually building and installing the extension. The Dockerfile also includes full build steps and dependencies. I think it is worth to add a Dockerfile for this project.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+RUN echo "===> Adding prerequisites..."                      && \
+    apt-get update -y                                        && \
+    DEBIAN_FRONTEND=noninteractive                              \
+        apt-get install --no-install-recommends --allow-downgrades -y -q \
+                ca-certificates                                 \
+                build-essential git                             \
+                postgresql-server-dev-9.6                       \
+                libpq-dev=$PG_MAJOR\* libpq5=$PG_MAJOR\* libecpg-dev && \
+    echo "===> Building datasketches..."                     && \
+    git clone https://github.com/apache/incubator-datasketches-cpp.git datasketches-cpp && \
```
Comment (jmalkin) [inline Dockerfile]: I'm pretty sure Apache policy means we'll need to pin this to an official release.

Comment (phstudy) [reply Dockerfile]: fixed

**Comment by AlexanderSaydakov** [comment]
why PostgreSQL 9.6? can this be overridden easily?

**Comment by phstudy** [comment]
> why PostgreSQL 9.6? can this be overridden easily?

PostgreSQL 9.6 is just default version. I have updated the PR to make version overridable.

Thanks.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+ARG PG_MAJOR_VERSION=9.6
+ARG PG_MINOR_VERSION=9.6
```
Comment (AlexanderSaydakov) [inline Dockerfile]: how are these used? why we need them separately? isn't it like 9 - major, 6 - minor?

Comment (phstudy) [reply Dockerfile]: [PostgreSQL version policy](https://www.postgresql.org/support/versioning/) changed after 9.6, so 9.6 is also major version.

~~In order to match distribution version with docker base image, we have to separate the version.~~

Comment (phstudy) [reply Dockerfile]: ~~I have updated the default `PG_MINOR_VERSION` and `BASE_IMAGE_VERSION` to an accurate version. This should make it more easy to understand the purpose of those versions.~~

Comment (phstudy) [reply Dockerfile]: ~~`PG_MAJOR_VERSION` is part of package name, for example: [`postgresql-server-dev-10`]https://packages.ubuntu.com/bionic/postgresql-server-dev-10) and `postgresql-server-dev-9.6`.
`PG_MINOR_VERSION` is specified version of package, for example: `postgresql-server-dev-10=10.13-1\*` and `postgresql-server-dev-10=9.6.18\*`.
`BASE_IMAGE_VERSION` is base image version.~~

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    DEBIAN_FRONTEND=noninteractive                              \
+        apt-get install --no-install-recommends --allow-downgrades -y -q \
+                ca-certificates                                 \
+                build-essential wget                            \
+                postgresql-server-dev-$PG_MAJOR=$PG_MINOR\*     \
+                libpq-dev=$PG_MINOR\* libpq5=$PG_MINOR\* libecpg-dev && \
+    \
+    \                
+    echo "===> Building datasketches..."                     && \
+    wget https://github.com/apache/incubator-datasketches-cpp/archive/$DATASKETCHES_CPP.tar.gz && \
```
Comment (leerho) [inline Dockerfile]: Downloading apache releases from github is not allowed.   And to do it automatically requires a script to balance between mirrors.  Please read [](https://infra.apache.org/release-download-pages).

Comment (leerho) [reply Dockerfile]: [https://infra.apache.org/release-download-pages](https://infra.apache.org/release-download-pages)

Comment (phstudy) [reply Dockerfile]: I borrow the download code from [apache/fluo](https://github.com/apache/fluo-docker/blob/1ccfc40d652fc715dee33a563606014356cff595/Dockerfile#L40). It should fit the downloading policy.

Comment (leerho) [reply Dockerfile]: Not necessarily.   I would post this question on general@incubator.apache.org and see what the response is.  I have been told by very senior folks at ASF that there are many older Apache projects that violate a lot of ASF policies but they have been grandfathered in.  Also the policies themselves have changed over the years as ASF has grown.  So the best advice would be to ask some of the senior folks like Justin Mclean, Dave Fischer, etc.

**Comment by phstudy** [comment]
The base image (debian buster) only supports LLVM 7.0 which failed to build extensions for postgresql 11+.

**Comment by phstudy** [comment]
> The base image (debian buster) only supports LLVM 7.0 which failed to build extension for postgresql 11+.

Fixed. Adding the following code in Makefile:
```
COMPILE.cxx.bc = $(CLANG) -xc++ -Wno-ignored-attributes $(BITCODE_CXXFLAGS) $(CPPFLAGS) -emit-llvm -c

%.bc : %.cpp
	$(COMPILE.cxx.bc) -o $@ $<
	$(LLVM_BINPATH)/opt -module-summary -f $@ -o $@
```

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+ARG PG_MAJOR_VERSION=12
+ARG PG_MINOR_VERSION=12.4
+ARG BASE_IMAGE_VERSION=12.4
```
Comment (phstudy) [inline Dockerfile]: The base image (debian buster) only supports LLVM 7.0 which failed to build extension for postgresql 11+.

Comment (phstudy) [reply Dockerfile]: Fixed. Adding the following code to Makefile:
```
COMPILE.cxx.bc = $(CLANG) -xc++ -Wno-ignored-attributes $(BITCODE_CXXFLAGS) $(CPPFLAGS) -emit-llvm -c

%.bc : %.cpp
	$(COMPILE.cxx.bc) -o $@ $<
	$(LLVM_BINPATH)/opt -module-summary -f $@ -o $@
```

This patch should also fix #25.

File: Dockerfile (hunk 2 above)
Comment (phstudy) [reply Dockerfile]: I have made some update:

1. Default base image version was changed from `9.6` to `latest`.

2. `PG_MAJOR_VERSION` & `PG_MINOR_VERSION` arguments were removed and I parse them from pre-installed packages.

3. Only `BASE_IMAGE_VERSION` argument is remained and it is overridable, example:
  ```
docker build --pull --build-arg BASE_IMAGE_VERSION=10 -t datasketch-postgres:10 .
```

4. Add continuous integration with Docker on Travis CI

File: Makefile [hunk 5]
```diff
@@ -43,6 +43,13 @@ PG_CONFIG = pg_config
 PGXS := $(shell $(PG_CONFIG) --pgxs)
 include $(PGXS)
 
+# fix LLVM JIT compilation error
+COMPILE.cxx.bc = $(CLANG) -xc++ -Wno-ignored-attributes $(BITCODE_CXXFLAGS) $(CPPFLAGS) -emit-llvm -c
```
Comment (AlexanderSaydakov) [inline Makefile]: will this work with gcc?

Comment (phstudy) [reply Makefile]: I have added llvm check before applying the patch.

```
ifeq ($(with_llvm), yes)
	COMPILE.cxx.bc = $(CLANG) -xc++ -Wno-ignored-attributes $(BITCODE_CXXFLAGS) $(CPPFLAGS) -emit-llvm -c
endif


%.bc : %.cpp
	$(COMPILE.cxx.bc) -o $@ $<
	if [ "$(with_llvm)" = "yes" ]; then $(LLVM_BINPATH)/opt -module-summary -f $@ -o $@; fi
```

Thanks.

**Pull request merged**